Add JSON schema based validation to mbed config script

scartmell-arm opened the pull request:
## Description
The current configuration system allows a user to specify various compile time configuration options for various targets within JSON files in the projects directory mbed_app.json for applications, and mbed_lib.json for libraries. The parameters will be converted to a collection of macros that are defined when compiling the code. The format of the JSON file is defined [here.](https://github.com/ARMmbed/Handbook/blob/5.5/docs/advanced/config_system.md)

Currently, the process of the checking the validity of the config file is done through a series of manual checks in `tools/config/__init__.py`. This PR will replace the manual checks with a solution that uses JSON schema to validate the configuration files.


## Status
**IN DEVELOPMENT**

## Todos
- [x] Define the JSON schema for app and lib library files
- [x] Remove remaining manual checks

## Steps to test or reproduce
Test the current validation test cases still all pass:

Linux: `$ PYTHONPATH=. python -m pytest tools/test/config/config_test.py`

Windows: `$ set PYTHONPATH=. && python -m pytest tools/test/config/config_test.py`

Commit:
Add JSON schema based validation to mbed config script

- Added app and lib JSON schema definition files which specify the valid
  keys and values that can be used in mbed library and application
  configuration files. The primary different between the app and lib
  schema is that the lib config requires a name key.

- Modified the expected error code in some of the test cases. The error
  message is now issued by the JSON validator.

- Added some validation code to the config script which checks the
  validity of the mbed_app.json file when it is initially loaded.

- Added some validation code to config script which checks each of the
  mbed_lib.json scripts when they are loaded.

- Removed manual checks for allowable config keys from within the mbed_app
  and mbed_lib json files.

- Removed the check_dict_types() function which was no longer being
  called.

theotherjimmy reviewed (approved):
Looks great so far! In it's current form, it's a huge improvement over the old validation mechanism. Be aware that #4984 may create a conflict with the test changes.

Commit:
Add JSON Schema library to requirements.txt

Commit:
Remove config_path setting and amend regex in mbed config JSON schema

- Removed the config_path setting from schema and moved the addition of the
config_path value after the validation is done.

- Altered the macro validation regex to be more lenient. Now verifies that
  if '=' is used in the macro definition that something comes after it.

Commit:
Refactor common JSON schema definitions into a seperate file

- Moved all common definitions from schema_app and schema_lib into a
  separate definitions file.

- Changed the calling code to resolve multiple schema files correctly.

theotherjimmy reviewed (commented):
Was this an omission?

Commit:
Remove placeholder $id from JSON schema and fix description

Commit:
Fix test errors produced by JSON schema after rebase

- Modified the expected error code in some of the test cases. The error
  message is now issued by the JSON validator.

- Stop validation code being called if no application configuration is
  given to validate.

- Change test mock code to check for specific calls, instead of number of
  calls.

- Derive absolute paths for JSON schema files before loading them, which
  seems to work better with the URI path resolution of the schema parser.

theotherjimmy reviewed (changes requested):
I think the schema can be more restrictive, and therefore more helpful at pointing out errors.

theotherjimmy reviewed (approved):
Let's get this in, and I'll submit another PR to make it more strict. @cmonr If you could do the honors
